### ISSUE TITLE
Decoding speed bug introduced

### ISSUE THREAD

**Issue by pps83** [powturbo/TurboPFor#56]
Just to make sure it's tracked properly, decoding speed reduced by more than 10% for p4nzdec256v32 (my personal project seems to have best results with this function).

[9dad490](https://github.com/powturbo/TurboPFor/commit/9dad4903bc76bcac1f3fe40c2b8b2f42526f118b) doesn't have the problem, [5dff9d3f](https://github.com/powturbo/TurboPFor/commit/5dff9d3fc120fafa13cedbc8ea0f249e8291215a) has the problem.
It's absolutely clearly almost 10% slower for entire test where p4nzdec256v32 connsumes perhaps 25% of the cpu time of the entire test.

Times of my tests with these TurboPFor versions:
9dad490 - **16.605sec** ~MyAlgo
profiler data:
![image](https://user-images.githubusercontent.com/1614246/71307177-5c16ed00-239f-11ea-9cfb-92f4a9866a2d.png)

p4nzdec256v32 roughly takes 3.560sec (21.44% of 16.605sec)

5dff9d3f - **17.636sec** ~MyAlgo
profiler data:
![image](https://user-images.githubusercontent.com/1614246/71307179-646f2800-239f-11ea-86be-712f7de56f4a.png)

p4nzdec256v32 roughly takes 4.213sec (23.89% of 17.636sec)
These tests are performed with absolutely identical code, simply TurboPFor version changed.
Also, the test uses 3 threads (according to the debugger, so, actual performance hit in p4nzdec256v32 seems to be way more than 10%, and might easily be in the range of 20%-30%). Also, times were measured without profiler (just to make it clear, it was fully optimized runs of 64-bit builds of TurboPFor)

Unfortunately, due to the way TurboPFor is managed, it's in permanently broken state and I wasn't able to bisect to find offending commit: if I try to check out anything in between, it's always broken and doesn't build with all kinds of compilation errors. I'd strongly recommend adjusting your approach: it's not suitable for projects that are worked on and used by more than one person.

**Comment by powturbo** [comment]
I've benchmarked the current version with the old version before the changes (convert macros to inline functions in bitutil.h), but the speed is the same.  
This is also valable for all avx2 "p4n?enc/dec" and "bitn?pack/unpack" functions.

The benchmarks are done with a very stable skylake cpu (not overclocked) using ubuntu 19.10 and
gcc 9.2 for both versions.

                            ./icapp -e12 ZIPF -I15 -J15
<pre>
 1458.88     851790  21.29%    7285.21 p4nzenc256v32    TurboPFor256 zigzag  NEW version
 1475.47     851790  21.29%    7246.38 p4nzenc256v32    TurboPFor256 zigzag  OLD version
</pre>
It has probably something to do with the microsoft compiler.
I've only avx2 notebooks for testing (no avx2 desktop cpu) under windows.

Have you also done the tests with the current version including the project files from github?
Could you please, make the tests with icapp?

Edit: I've uploaded a new [bitunpack.c](https://github.com/powturbo/TurboPFor/blob/master/bitunpack.c#L533) with the macros enabled for the ms compiler.

**Comment by pps83** [comment]
But current master branch doesn't build with MS compiler, also due to removal of those _sse/_avx files, I'm not even sure how to build it properly with visual studio.

**Comment by pps83** [comment]
I just tried latest master, and I confirm that the problem is fixed now. Also, it all builds as-is now and I didn't need to fix/patch anything.

**Comment by powturbo** [comment]
can you please, try to disable the macros in bitunpack.c line [533 ](https://github.com/powturbo/TurboPFor/blob/master/bitunpack.c#L533) (change it to #if 0 )
and compare the results. We can see, if the ms compiler doesn't optimize well the "inline functions".

**Comment by pps83** [comment]
Just tried it, and yes, this results in the slow down. In general MS compilers are pretty good at optimizing, not sure what's going on, the only way to know what's happening is to compile with asm generation to inspect generated code.

By the way, it seems that somehwere you corrupted the project. It works only in 64 bit, but x86 build is broken now with tons of warnings and linker errors that I've mentioned elsewhere: these are x86 specific now:

```1>------ Rebuild All started: Project: TurboPFor, Configuration: Release Win32 ------
1>bitpack.c
1>bitunpack.c
1>bitutil.c
1>fp.c
1>transpose.c
1>trlec.c
1>trled.c
1>v8.c
1>vint.c
1>vp4c.c
1>vp4d.c
1>vsimple.c
1>bitpack_avx2.c
1>bitpack_sse.c
1>bitunpack_avx2.c
1>bitunpack_sse.c
1>transpose_avx2.c
1>transpose_sse.c
1>vp4c_avx2.c
1>vp4c_sse.c
1>vp4d_avx2.c
1>vp4d_sse.c
1>vp4d_sse.obj : warning LNK4006: _p4ndec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4nddec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4nd1dec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4nzdec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4dec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4ddec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4d1dec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4zdec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4c_sse.obj : warning LNK4006: _p4nenc256v32 already defined in vp4c_sse.obj; second definition ignored
1>vp4c_sse.obj : warning LNK4006: _p4ndenc256v32 already defined in vp4c_sse.obj; second definition ignored
1>vp4c_sse.obj : warning LNK4006: _p4nd1enc256v32 already defined in vp4c_sse.obj; second definition ignored
1>vp4c_sse.obj : warning LNK4006: _p4nzenc256v32 already defined in vp4c_sse.obj; second definition ignored
1>vp4c_sse.obj : warning LNK4006: _p4enc256v32 already defined in vp4c_sse.obj; second definition ignored
1>vp4c_sse.obj : warning LNK4006: _p4denc256v32 already defined in vp4c_sse.obj; second definition ignored
1>vp4c_sse.obj : warning LNK4006: _p4d1enc256v32 already defined in vp4c_sse.obj; second definition ignored
1>vp4c_sse.obj : warning LNK4006: _p4zenc256v32 already defined in vp4c_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tpenc256v2 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tpdec256v2 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tpenc256v4 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tpdec256v4 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tpenc256v8 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tpdec256v8 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tp4enc256v2 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tp4dec256v2 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tp4enc256v4 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tp4dec256v4 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tp4enc256v8 already defined in transpose_sse.obj; second definition ignored
1>transpose_sse.obj : warning LNK4006: _tp4dec256v8 already defined in transpose_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitnunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitndunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitnd1unpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitnzunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitnfunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitzunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitdunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitd1unpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitfunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: _bitf1unpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: __bitunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: __bitdunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: __bitd1unpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitunpack_sse.obj : warning LNK4006: __bitzunpack256v32 already defined in bitunpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitnpack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitndpack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitnd1pack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitnzpack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitnfpack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitpack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitdpack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitd1pack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitfpack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitf1pack256v32 already defined in bitpack_sse.obj; second definition ignored
1>bitpack_sse.obj : warning LNK4006: _bitzpack256v32 already defined in bitpack_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4ndec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4nddec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4nd1dec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4nzdec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4dec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4ddec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4d1dec256v32 already defined in vp4d_sse.obj; second definition ignored
1>vp4d_sse.obj : warning LNK4006: _p4zdec256v32 already defined in vp4d_sse.obj; second definition ignored
1>TurboPFor.vcxproj -> C:\work-pps\TurboPFor\vs\vs2017\msvc.build\Win32-Release\TurboPFor.lib
1>Done building project "TurboPFor.vcxproj".
2>------ Rebuild All started: Project: icapp, Configuration: Release Win32 ------
2>icapp.c
2>getopt.c
2>Generating Code...
2>icapp.obj : error LNK2019: unresolved external symbol _p4ndec8 referenced in function _bench8
2>icapp.obj : error LNK2019: unresolved external symbol _p4ndec16 referenced in function _bench16
2>icapp.obj : error LNK2019: unresolved external symbol _p4ndec32 referenced in function _bench32
2>icapp.obj : error LNK2019: unresolved external symbol _p4ndec64 referenced in function _bench64
2>icapp.obj : error LNK2019: unresolved external symbol _p4nddec8 referenced in function _bench8
2>icapp.obj : error LNK2019: unresolved external symbol _p4nddec16 referenced in function _bench16
2>icapp.obj : error LNK2019: unresolved external symbol _p4nddec32 referenced in function _bench32
2>icapp.obj : error LNK2019: unresolved external symbol _p4nddec64 referenced in function _bench64
2>icapp.obj : error LNK2019: unresolved external symbol _p4nd1dec8 referenced in function _bench8
2>icapp.obj : error LNK2019: unresolved external symbol _p4nd1dec16 referenced in function _bench16
2>icapp.obj : error LNK2019: unresolved external symbol _p4nd1dec32 referenced in function _bench32
2>icapp.obj : error LNK2019: unresolved external symbol _p4nd1dec64 referenced in function _bench64
2>icapp.obj : error LNK2019: unresolved external symbol _p4nzdec8 referenced in function _bench8
2>icapp.obj : error LNK2019: unresolved external symbol _p4nzdec16 referenced in function _bench16
2>icapp.obj : error LNK2019: unresolved external symbol _p4nzdec32 referenced in function _bench32
2>icapp.obj : error LNK2019: unresolved external symbol _p4nzdec64 referenced in function _bench64
2>TurboPFor.lib(vp4c.obj) : error LNK2019: unresolved external symbol _p4dec16 referenced in function _p4sdec16
2>TurboPFor.lib(fp.obj) : error LNK2001: unresolved external symbol _p4dec16
2>TurboPFor.lib(vp4c.obj) : error LNK2019: unresolved external symbol _p4dec32 referenced in function _p4sdec32
2>TurboPFor.lib(fp.obj) : error LNK2001: unresolved external symbol _p4dec32
2>TurboPFor.lib(vp4c.obj) : error LNK2019: unresolved external symbol _p4dec64 referenced in function _p4sdec64
2>TurboPFor.lib(fp.obj) : error LNK2001: unresolved external symbol _p4dec64
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitpack128v16 referenced in function _v8nenc128v16
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitdpack128v16 referenced in function _v8ndenc128v16
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitd1pack128v16 referenced in function _v8nd1enc128v16
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitzpack128v16 referenced in function _v8nzenc128v16
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitpack128v32 referenced in function _v8nenc128v32
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitdpack128v32 referenced in function _v8ndenc128v32
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitd1pack128v32 referenced in function _v8nd1enc128v32
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitzpack128v32 referenced in function _v8nzenc128v32
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitunpack128v16 referenced in function _v8ndec128v16
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitzunpack128v16 referenced in function _v8nzdec128v16
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitdunpack128v16 referenced in function _v8nddec128v16
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitd1unpack128v16 referenced in function _v8nd1dec128v16
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitunpack128v32 referenced in function _v8ndec128v32
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitzunpack128v32 referenced in function _v8nzdec128v32
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitdunpack128v32 referenced in function _v8nddec128v32
2>TurboPFor.lib(v8.obj) : error LNK2019: unresolved external symbol _bitd1unpack128v32 referenced in function _v8nd1dec128v32
2>TurboPFor.lib(fp.obj) : error LNK2019: unresolved external symbol _p4enc128v16 referenced in function _fp2dfcmenc16
2>TurboPFor.lib(fp.obj) : error LNK2019: unresolved external symbol _p4enc128v32 referenced in function _fp2dfcmenc32
2>TurboPFor.lib(fp.obj) : error LNK2019: unresolved external symbol _p4enc128v64 referenced in function _fp2dfcmenc64
2>TurboPFor.lib(fp.obj) : error LNK2019: unresolved external symbol _p4dec8 referenced in function _fp2dfcmdec8
2>TurboPFor.lib(fp.obj) : error LNK2019: unresolved external symbol _p4dec128v16 referenced in function _fp2dfcmdec16
2>TurboPFor.lib(fp.obj) : error LNK2019: unresolved external symbol _p4dec128v32 referenced in function _fp2dfcmdec32
2>TurboPFor.lib(fp.obj) : error LNK2019: unresolved external symbol _p4dec128v64 referenced in function _fp2dfcmdec64
2>C:\work-pps\TurboPFor\vs\vs2017\msvc.build\Win32-Release\icapp.exe : fatal error LNK1120: 42 unresolved externals
2>Done building project "icapp.vcxproj" -- FAILED.
========== Rebuild All: 1 succeeded, 1 failed, 0 skipped ==========
```

**Comment by pps83** [comment]
In case if you case and want to see, you can output asm this way:

![image](https://user-images.githubusercontent.com/1614246/71327715-f4df7100-2535-11ea-9bbd-b0862edf45b3.png)


You can set that to a single file, or entire project.

**Comment by powturbo** [comment]
Ok, thank you. I'll reintroduce the macros in bitutil.h and make a switch to choose between inline functions and macros depending on the compiler. I think the ms compiler is getting confused because of a lot of simd unrolling in bitpack/bitunpack. 
I've got also the linker errors by setting the instruction set option to sse. I've changed this to avx, then it worked without errors. I think that's the only thing, I've changed in the project files.
